Add CPP stacktrace on CUDA OOM

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82665

When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

rohan-varma committed:
CPP stacktrace on CUDA OOM

[ghstack-poisoned]

rohan-varma committed:
Update on "Add CPP stacktrace on CUDA OOM"


When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

[ghstack-poisoned]

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -52,58 +56,118 @@ static inline void PyErr_SetString(PyObject* type, const std::string& message) {
-    PyErr_SetString(e.python_type(), msg);                                   \
-    retstmnt;                                                                \
-  }
+#if defined(USE_CUDA)
```
I hate this. Can you instead move CUDAOutOfMemoryError so it is defined unconditionally and then eliminate the #if

rohan-varma replied:
Sure, it seems that we'd lose the C10_CUDA_API declaration of the error though, and instead just have it as a C10_API,  not sure what the difference is there.

ezyang replied:
it doesn't matter; the API decl has to match what library it's in but if you move it to libc10.so that's correct

rohan-varma committed:
Update on "Add CPP stacktrace on CUDA OOM"


When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add CPP stacktrace on CUDA OOM"


When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add CPP stacktrace on CUDA OOM"


When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add CPP stacktrace on CUDA OOM"


When reporting CUDA OOM error, defaults to showing the C++ stacktrace to help with debugability in terms of when this happened / which operator caused it. 

Had to make separate macros for handling the torch errors based on whether CUDA is available or not, because you cannot have a define directive within another directive: https://stackoverflow.com/questions/10074520/why-compiler-complain-about-this-macro-declaration. Feel free to suggest better alternatives!

[ghstack-poisoned]

zdevito commented on torch/csrc/Exceptions.h:
```diff
@@ -92,6 +92,11 @@ static inline void PyErr_SetString(PyObject* type, const std::string& message) {
   }                                                                          \
+  catch (const c10::CUDAOutOfMemoryError& e) {                               \
+    auto msg = e.what();                                                     \
+    PyErr_SetString(THPException_LinAlgError, torch::processErrorMsg(msg));  \
```
`THPException_LinAlgError` is probably not the right value.

rohan-varma replied:
oops this is copy paste vestige, it should probably be `PyExc_RuntimeError`.

zdevito commented on torch/csrc/Exceptions.h:
```diff
@@ -92,6 +92,11 @@ static inline void PyErr_SetString(PyObject* type, const std::string& message) {
+    retstmnt;                                                                \
+  }                                                                          \
   catch (const c10::Error& e) {                                              \
     auto msg = torch::get_cpp_stacktraces_enabled()                          \
```
Maybe we should just add a bool to c10::Error to indicate whether an error type should always include the stack and just put the logic here. CUDAOutOfMemoryError could then set the field to true in its constructor.

rohan-varma replied:
Filed https://github.com/pytorch/pytorch/issues/82751 to track this as follow-up work.

albanD commented:
Why is setting `TORCH_SHOW_CPP_STACKTRACES=1` not enough for this?

rohan-varma commented:
> Why is setting `TORCH_SHOW_CPP_STACKTRACES=1` not enough for this?

@albanD There is discussion internally about defaulting to showing c++ stacktraces on CUDA OOMs, as it helps with debugability in terms of narrowing down the actual operator causing the OOM.

ezyang reviewed (approved):
Zach's comments are on point, but otherwise I don't see any reason to block landing this PR

albanD commented:
> @albanD There is discussion internally about defaulting to showing c++ stacktraces on CUDA OOMs, as it helps with debugability in terms of narrowing down the actual operator causing the OOM.

But if you always want c++ stacktraces internally, why not just switch the default there to always show them (for OOM and other errors)?

ezyang commented:
This is a good point: for non interactive flows, we probably should turn on c++ stack traces by default. But I'm still happy to land this PR for cases where we didn't manage to fill it

albanD commented:
Note that a side effect of this PR is that now, all OSS users will see a giant c++ stack trace on OOM errors. Which will be quite confusing for most of them and they will have no way to silence it.

rohan-varma commented:
I think this needs further discussion given @albanD's point. This seems good for developers but harder to action for users, so maybe we should consider just defaulting to set this value for internal flows? 

@ezyang @albanD

albanD commented:
Making it the default as an opt-in (used internally) sounds good to me.
